packages/core/rsc/streamable.tsx

```diff
@@ -377,31 +356,55 @@ function createStreamableValueImpl<T = any, E = any>(initialValue?: T) {
     append(value: T) {
       assertStream('.append()');
 
-      if (
-        typeof currentValue !== 'string' &&
-        typeof currentValue !== 'undefined'
-      ) {
-        throw new Error(
-          `.append(): The current value is not a string. Received: ${typeof currentValue}`,
-        );
-      }
-      if (typeof value !== 'string') {
+      if (typeof currentValue === 'undefined') {
+        if (typeof value === 'string' || isValidElement(value)) {
+          currentPatchValue = undefined;
+          currentValue = value;
+        } else {
+          throw new Error(
+            `.append(): The value type can't be appended to the stream. Received: ${typeof value}`,
+          );
+        }
+      } else if (typeof currentValue === 'string') {
+        if (typeof value === 'string') {
+          currentPatchValue = [0, value];
+          (currentValue as string) = currentValue + value;
+        } else if (isValidElement(value)) {
+          currentPatchValue = [1, value];
+          (currentValue as unknown as ReactElement) = (
+            <>
+              {currentValue}
+              {value}
+            </>
+          );
+        } else {
+          throw new Error(
+            `.append(): The value type can't be appended to the stream. Received: ${typeof value}`,
+          );
+        }
+      } else if (isValidElement(currentValue)) {
+        if (typeof value === 'string' || isValidElement(value)) {
+          currentPatchValue = [1, value];
+          (currentValue as ReactElement) = (
+            <>
+              {currentValue}
+              {value}
+            </>
+          );
+        } else {
+          throw new Error(
+            `.append(): The value type can't be appended to the stream. Received: ${typeof value}`,
+          );
+        }
```

Comment:
the check for `!(string | validelement)` is duplicated 3 times. you could invert the if-else and pull it out

Reply:
Yeah part of it is to bypass TS checks as it's not that intelligent to handle the branching if `typeof value` is stored as a variable.